backend/push.js

```diff
@@ -1,53 +1,292 @@
-const giConfig = require('../giconf.json')
+import { aes128gcm } from '@apeleghq/rfc8188/encodings'
+import encrypt from '@apeleghq/rfc8188/encrypt'
+import sbp from '@sbp/sbp'
+import { PUBSUB_INSTANCE } from './instance-keys.js'
+import rfc8291Ikm from './rfc8291Ikm.js'
+import { getVapidPublicKey, vapidAuthorization } from './vapid.js'
+
+// const pushController = require('web-push')
 const { PUSH_SERVER_ACTION_TYPE, REQUEST_TYPE, createMessage } = require('../shared/pubsub.js')
 
-// NOTE: VAPID public/private keys can be generated via 'npx web-push generate-vapid-keys' command.
-const publicKey = process.env.VAPID_PUBLIC_KEY || giConfig.VAPID_PUBLIC_KEY
-const privateKey = process.env.VAPID_PRIVATE_KEY || giConfig.VAPID_PRIVATE_KEY
-pushController.setVapidDetails(
-  process.env.VAPID_EMAIL || giConfig.VAPID_EMAIL,
-  publicKey,
-  privateKey
-)
+const addSubscriptionToIndex = async (subcriptionId: string) => {
+  await sbp('okTurtles.eventQueue/queueEvent', 'update-webpush-indices', async () => {
+    const currentIndex = await sbp('chelonia/db/get', '_private_webpush_index')
+    // Add the current subscriptionId to the subscription index. Entries in the
+    // index are separated by \x00 (NUL). The index itself is used to know
+    // which entries to load.
+    const updatedIndex = `${currentIndex ? `${currentIndex}\x00` : ''}${subcriptionId}`
+    await sbp('chelonia/db/set', '_private_webpush_index', updatedIndex)
+  })
+}
+
+const deleteSubscriptionFromIndex = async (subcriptionId: string) => {
+  await sbp('okTurtles.eventQueue/queueEvent', 'update-webpush-indices', async () => {
+    const currentIndex = await sbp('chelonia/db/get', '_private_webpush_index')
+    const index = currentIndex.indexOf(subcriptionId)
+    if (index === -1) return
+    const updatedIndex = currentIndex.slice(0, index > 1 ? index - 1 : 0) + currentIndex.slice(index + subcriptionId.length)
+    await sbp('chelonia/db/set', '_private_webpush_index', updatedIndex)
+  })
+}
+
+const saveSubscription = (server, subscriptionId) => {
+  sbp('chelonia/db/set', `_private_webpush_${subscriptionId}`, JSON.stringify({
+    subscription: server.pushSubscriptions[subscriptionId],
+    channelIDs: [...server.pushSubscriptions[subscriptionId].subscriptions]
+  })).catch(e => {
+    console.error(e, 'Error saving subscription', subscriptionId)
+  })
+}
+
+export const addChannelToSubscription = (server: Object, subscriptionId: string, channelID: string): void => {
+  server.pushSubscriptions[subscriptionId].subscriptions.add(channelID)
+  saveSubscription(server, subscriptionId)
+}
+
+export const deleteChannelFromSubscription = (server: Object, subscriptionId: string, channelID: string): void => {
+  server.pushSubscriptions[subscriptionId].subscriptions.delete(channelID)
+  saveSubscription(server, subscriptionId)
+}
+
+// Generate an UUID from a `PushSubscription'
+export const getSubscriptionId = async (subscriptionInfo: Object): Promise<string> => {
```

Comment:
What's the reason you decided to create this type of UUID instead of just concatenating the 3 values and pushing them through blake32hash?

Reply:
No strong reason. It seemed more apt as previously we were using the SW client id (which is a UUID) for a similar purpose.